blog: goose janitor (#4131)
block/goose: documentation/blog/2025-08-28-ai-teammate/index.md

authors: 
    - angie
---

![blog banner](goose-janitor.png)

Over the years, Block's iOS engineers have felt the pain of technical debt accumulating. Feature flags are a specific example of this. Even when successfully rolled out, they tend to linger in the codebase, each one a small weight slowing down development.

In the early stages of 2025, with a renewed focus on developer acceleration, the Foundation iOS team decided to organize 'Feature Flag Removal Month' - an opportunity for teams in the large iOS monorepo to come together and delete _possibly hundreds of thousands of lines of code of dead code_. 

With the serendipitous launch of Goose recipes around the same time, the team wondered could a dedicated recipe aid this effort? [Gemma Barlow](https://www.linkedin.com/in/gemmakbarlow/), a newer iOS engineer on the team, wanted to find out. 

<!-- truncate -->

## Phase 1: Making Knowledge AI-Accessible

Gemma's first move was to utilize an existing system of `.mdc`, `.goosehints` and other symlinked context documents to capture how to safely remove feature flags from the repository.

> She created documentation that allows AI agents to glean enough context to perform their work accurately.

Reviewer: ```suggestion
She added documentation that allows AI agents to glean enough context to perform their removal work accurately.
```

## What's Next?

The success of Goose Janitor opens up fascinating possibilities. What other forms of technical debt could benefit from this approach? What other "AI team members" could we build to handle routine but knowledge intensive work?

As we move toward an AI-first future, stories like Gemma's show us the path. Not just using AI tools, but thinking systematically about how to make our codebases and processes AI-ready.

The future of software development is mixed teams where AI agents are autonomous contributors, handling the maintenance work that keeps our systems healthy while humans focus on building the future.

---

Want the tweak the base recipe for your own needs? Check out [Clean up feature flag](/recipes/detail/?id=clean-up-feature-flag) in our Recipe Cookbook!

<head>
  <meta property="og:title" content="When AI Becomes Your New Team Member: The Story of Goose Janitor" />
  <meta property="og:type" content="article" />
  <meta property="og:url" content="https://block.github.io/goose/blog/2025/08/28/ai-teammate" />

>   <meta property="og:description" content="How one iOS developer turned a Slack conversation into autonomous AI-powered technical debt cleanup" />

Reviewer: ```suggestion
  <meta property="og:description" content="How an engineering team is trialling autonomous AI-powered technical debt cleanup" />
```

  <meta property="og:image" content="https://block.github.io/goose/assets/images/goose-janitor-129889884d9265d001fe12cbfde03d57.png" />
  <meta name="twitter:card" content="summary_large_image" />
  <meta property="twitter:domain" content="block.github.io/goose" />
  <meta name="twitter:title" content="When AI Becomes Your New Team Member: The Story of Goose Janitor" />

>   <meta name="twitter:description" content="How one iOS developer turned a Slack conversation into autonomous AI-powered technical debt cleanup" />

Reviewer: ```suggestion
  <meta name="twitter:description" content="How an engineering team is trialling autonomous AI-powered technical debt cleanup" />
```
